### ISSUE TITLE
Add Subresource Integrity support for ModuleFederation

### ISSUE THREAD

**Issue by amitbet** [webpack/webpack#14310]
## Feature request

**What is the expected behavior?**
I am working in a large enterprise org, and we are using ModuleFederation for our MicroFrontEnds implementation.
I would like to have the ability to use SRI (via plugin or some flag in the ModFed settings)

**What is motivation or use case for adding/changing the behavior?**
The dynamic loading of JS code is an opening for supply chain attack, this concern was raised by our security team and we need some mechanism to ensure the integrity of the loaded JS.

**How should this be implemented in your opinion?**
the createElement in LoadScriptRuntimeModule.js should be modified to enable adding the "integrity=[contentHash]" attribute, and it should be flaggable in the config.

if there's a better way i would like to hear about it (there is no documentation online about SRI and ModFed)

**Are you willing to work on this yourself?**
yes, but I would need some help in order to "do no harm"

**Comment by alexander-akait** [comment]
Anyway you can do it using plugin

**Comment by amitbet** [comment]
how? the SubResource integrity plugin doesn't work with federated modules as far as I can see...

**Comment by sokra** [comment]
Where should we get the integrity from?
The idea with module federation is that the remote script can change independent from the consuming script. So it can't be passed via config. It need to be provided at runtime.

I would recommend to opt-out of automatic remote script loading and let your server compose a html with remote script tags including integrity.

**Comment by amitbet** [comment]
The chunk integrity can be hardcoded in the remote-entry.js file, for the top level, there are several options depending on what you are implementing, if there is a declaration of the remotes, you can include the integrity hash in that definition, if it is more dynamic than that (as it is in our case as well),
the top level integrity should be coming from the same code that knows which concrete module you will be consuming (in our app we resolve a tree of modules with semver queries into concrete versions, that can have a hash associated with each).

**in summary:** in the more dynamic cases, you should leave the top level open to receive the integrity from outside.

**Comment by aversini** [comment]
I am in the same boat as you @amitbet : big corporation with need for both module federation and SRI. Did you figure out a way? I’m wondering if we could simply limit SRI to app level, and ignore it for chunks coming from the remotes? We are using the `webpack-subresource-integrity` plugin but it doesn’t seem to being able to let us bypass some chunks - it’s all or nothing…

**Comment by jaecktec** [comment]
I just stumbled upon this, and I have a hypothetical solution. 

https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes

using this guide you could resolve the url from some sort of rest-api (json/http) which responds with the latest file-name and it's hash. The script would then create the <script> element with that hash in mind. 

Any thoughts on that?

**Comment by subeshb1** [comment]
Any updates on this?

**Comment by alexander-akait** [comment]
bump

**Comment by azinod** [comment]
+1 to this

**Comment by sowtame** [comment]
bump

**Comment by fdecollibus** [comment]
+1

**Comment by AmsterGet** [comment]
+1

**Comment by sowtame** [comment]
hi, I made a temporary package [@sowtame/webpack-subresource-integrity](https://github.com/waysact/webpack-subresource-integrity/pull/220), supports wmf.
it's created from this [pr](https://github.com/waysact/webpack-subresource-integrity/pull/220)

**Comment by noahnu** [comment]
This wasn't actually resolved, was it? That's just the stale bot closing the issue?

**Comment by alexander-akait** [comment]
It is a plugin side, there is a PR https://github.com/waysact/webpack-subresource-integrity/pull/220
